app/com/gu/floodgate/reindex/ProgressTracker.scala

```diff
@@ -0,0 +1,127 @@
+  case class UpdateProgress(result: Try[WSResponse], contentSource: ContentSource, runningJob: RunningJob)
+  case class Cancel(contentSource: ContentSource, runningJob: RunningJob)
+
+}
+
+class ProgressTracker(ws: WSAPI, runningJobService: RunningJobService, jobHistoryService: JobHistoryService) extends Actor with ActorLogging with StrictLogging {
+
+  import context.become
+
+  final def receive = sleeping
+
+  private var nextPollSchedule: Option[Cancellable] = None
+
+  private def sleeping: Receive = {
+    case TrackProgress(contentSource, runningJob) =>
+      askForProgress(contentSource, runningJob)
+      become(waitingForProgress)
+
+    case Cancel(contentSource, runningJob) => completeProgressTracking("cancelled", contentSource, runningJob)
+    case other => logger.warn(s"Unexpected message received while sleeping: $other")
+  }
+
+  private def waitingForProgress: Receive = {
+    case UpdateProgress(result, contentSource, runningJob) =>
+      updateProgress(result, contentSource, runningJob)
+      become(sleeping) // no matter whether the result is good or bad, we should go back to sleeping
+
+    case Cancel => become(waitingForFinalProgress) // don't stop ourselves until we have received the final CAPI response (just to avoid a dead letter)
+    case other => logger.warn(s"Unexpected message received while waiting for response: $other")
+  }
+
+  private def waitingForFinalProgress: Receive = {
+    case UpdateProgress(_, contentSource, runningJob) =>
+      completeProgressTracking("cancelled", contentSource, runningJob) // we've received our last message, so now we can stop (we ignore the response)
+    case Cancel => log.debug("Ignoring a Cancel message because we're already planning to stop after we receive the final response")
+    case other => logger.warn(s"Unexpected message received while waiting for final response: $other")
+  }
+
+  private def askForProgress(contentSource: ContentSource, runningJob: RunningJob): Unit = {
+    val myself = self // make this a val to fix its value, because the onComplete will occur later in a different thread
+    ws.url(contentSource.reindexEndpoint).get() onComplete { result => myself ! UpdateProgress(result, contentSource, runningJob) }
+  }
+
+  private def updateProgress(result: Try[WSResponse], contentSource: ContentSource, runningJob: RunningJob): Unit = {
+    result match {
+      case Success(response) =>
+        response.status match {
+          case 200 => onSuccess(response, contentSource, runningJob)
+          case _ => onFailure(contentSource, runningJob)
+        }
+      case Failure(e) => onFailure(contentSource, runningJob)
+    }
+  }
+
+  private def onSuccess(response: WSResponse, contentSource: ContentSource, runningJob: RunningJob): Unit = {
+    response.json.validate[Progress].fold(
+      error => {
+        logger.warn(s"Content source with id: ${contentSource.id} appears to be returning progress updates in an incorrect format. Cancelling reindex.")
```

Comment:
Not strictly correct. We don't cancel the reindex, we just mark it as cancelled and stop monitoring it.